### PR TITLE
feature: add route rule support

### LINKED ISSUE #42: Need support of route rule
Provide information similar to `ip rule`.

This task is complex as the `rust-netlink` project does not support querying route rule yet.

### PR DESCRIPTION
**Pull request by ffmancera**
Example output for `npc rule`.

```
[
  {
    "action": "Table",
    "address_family": "IPv4",
    "flags": 0,
    "table": 254,
    "tos": 16,
    "dst": "192.0.2.2/32",
    "src": "192.0.2.1/32",
    "iif": "eth1",
    "oif": "eth2",
    "priority": 32762,
    "suppress_prefix_len": 4294967295,
    "protocol": "Kernel"
  },
]
```

Signed-off-by: Fernando Fernandez Mancera <ffmancera@riseup.net>

### CODE REVIEW

**Comment by ffmancera** [comment]
This depends on https://github.com/little-dude/netlink/pull/101

**Comment by ffmancera** [comment]
Ready for review.

File: src/lib/route_rule.rs [hunk 1]
```diff
@@ -0,0 +1,206 @@
...
+#[derive(Serialize, Deserialize, Debug, PartialEq, Clone)]
+pub enum RuleAction {
+    /* Pass to fixed table */
+    Table,
+    /* Jump to another rule */
+    Goto,
+    /* No operation */
+    Nop,
+    Res3,
+    Res4,
```
Comment (cathay4t) [inline src/lib/route_rule.rs]: Unless you know what is `Res3` and `Res4` is, please remove it.

There is not document in kernel mentioned this. And iproute2 has not such code neither.

File: src/lib/route_rule.rs [hunk 2]
```diff
@@ -0,0 +1,206 @@
...
+        Self::Unknown
+    }
+}
+
+#[derive(Serialize, Deserialize, Debug, PartialEq, Clone, Default)]
+pub struct RouteRule {
+    pub action: RuleAction,
+    pub address_family: AddressFamily,
+    pub flags: u32,
+    pub table: u32,
```
Comment (cathay4t) [inline src/lib/route_rule.rs]: We should hide `table` when it is `RT_TABLE_UNSPEC`

File: src/lib/route_rule.rs [hunk 3]
```diff
@@ -0,0 +1,206 @@
...
+        rules.push(get_rule(rt_msg)?);
+    }
+    let mut links = handle.rule().get(IpVersion::V4).execute();
+    while let Some(rt_msg) = links.try_next().await? {
+        rules.push(get_rule(rt_msg)?);
+    }
+    Ok(rules)
+}
+
+fn get_rule(rule_msg: RuleMessage) -> Result<RouteRule, NisporError> {
```
Comment (cathay4t) [inline src/lib/route_rule.rs:126]: This function never raise `NisporError`, no need to use `Result<>`.

**Review by cathay4t — CHANGES_REQUESTED** [review]

File: src/lib/route_rule.rs [hunk 4]
```diff
@@ -0,0 +1,206 @@
...
+    #[serde(skip_serializing_if = "Option::is_none")]
+    pub mask: Option<u32>,
+    #[serde(skip_serializing_if = "Option::is_none")]
+    pub flow: Option<u32>,
+    #[serde(skip_serializing_if = "Option::is_none")]
+    pub tun_id: Option<u32>,
+    #[serde(skip_serializing_if = "Option::is_none")]
+    pub suppress_ifgroup: Option<u32>,
+    #[serde(skip_serializing_if = "Option::is_none")]
+    pub suppress_prefix_len: Option<u32>,
```
Comment (cathay4t) [inline src/lib/route_rule.rs:95]: Both `suppress_prefix_len` and `suppress_ifgroup` should be hidded when equal to`std::u32::MAX`.

File: src/lib/route_rule.rs (hunk 1 above)
Comment (ffmancera) [reply src/lib/route_rule.rs]: Done, removed.

**Comment by cathay4t** [comment]
Please also include changed to `test_env`, so that I can verify it works manually.

File: src/lib/route_rule.rs (hunk 2 above)
Comment (ffmancera) [reply src/lib/route_rule.rs]: Done. Thanks!

File: src/lib/route_rule.rs (hunk 3 above)
Comment (ffmancera) [reply src/lib/route_rule.rs:126]: Done, thank you!

File: src/lib/route_rule.rs (hunk 4 above)
Comment (ffmancera) [reply src/lib/route_rule.rs:95]: Done, thank you!

**Comment by ffmancera** [comment]
> Please also include changed to `test_env`, so that I can verify it works manually.

Done, thank you!

File: src/lib/route_rule.rs [hunk 5]
```diff
@@ -0,0 +1,211 @@
...
+    }
+
+    rl
+}
+
+fn _addr_to_string(data: &[u8], family: &AddressFamily) -> String {
+    match family {
+        AddressFamily::IPv4 => parse_as_ipv4(data),
+        AddressFamily::IPv6 => parse_as_ipv6(data),
+        _ => format!("{:?}", data),
```
Comment (cathay4t) [inline src/lib/route_rule.rs]: Instead of putting junk string there, please use `eprintln!()` instead.

Comment (ffmancera) [reply src/lib/route_rule.rs]: Handeling error now.

**Review by cathay4t — APPROVED** [review]

**Pull request merged**